src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.sysds.common.Types;
+import org.apache.sysds.common.Types.DataType;
+import org.apache.sysds.conf.ConfigurationManager;
+import org.apache.sysds.runtime.DMLRuntimeException;
+import org.apache.sysds.runtime.io.IOUtilFunctions;
+import org.apache.sysds.runtime.meta.DataCharacteristics;
+
+public class ReaderFederated {
+    // private static final Log LOG = LogFactory.getLog(ReaderFederated.class.getName());
+
+    public static List<Pair<FederatedRange, FederatedData>> read(String fname, DataCharacteristics mc) {
+        List<Pair<FederatedRange, FederatedData>> res = new ArrayList<>();
+        try {
+            JobConf job = new JobConf(ConfigurationManager.getCachedJobConf());
+            Path path = new Path(fname);
+            FileSystem fs = IOUtilFunctions.getFileSystem(path, job);
+            FSDataInputStream data = fs.open(path);
+
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode json = mapper.readTree(data);
+
+            Iterator<Entry<String, JsonNode>> it = json.fields();
+            while(it.hasNext()) {
+                Entry<String, JsonNode> range = it.next();
+                FederatedData fedData = parseData(range.getValue());
+                FederatedRange fedRange = parseRange(range.getKey());
+
+                res.add(new ImmutablePair<>(fedRange, fedData));
+            }
+        }
+        catch(IOException e) {
+            fail("Unable to read federated matrix (" + fname + "): " + e.getMessage());
+        }
+        return res;
+    }
+
+    private static FederatedData parseData(JsonNode data) {
+
+        try {
+            DataType dt = Types.DataType.valueOf(data.get("dataType").asText());
+            String[] address = data.get("address").asText().split(":");
+            InetSocketAddress add = new InetSocketAddress(InetAddress.getByName(address[0]),
+                Integer.parseInt(address[1]));
+
+            return new FederatedData(dt, add, data.get("filepath").asText());
+        }
+        catch(NumberFormatException | UnknownHostException e) {
+            throw new DMLRuntimeException(e);
+        }
+    }
+
+    private static FederatedRange parseRange(String range) {
+        String[] split = range.split(" - ");
+        long[] start = parseStringArray(split[0]);
+        long[] end = parseStringArray(split[1]);
+
+        return new FederatedRange(start, end);
+    }
+
+    private static long[] parseStringArray(String arr) {
+        String[] vv = arr.replaceAll("\\[", "").replaceAll("\\]", "").replaceAll("\\s", "").split(",");
+        long[] res = new long[vv.length];
+        for(int i = 0; i < vv.length; i++) {
+            res[i] = Long.parseLong(vv[i]);
+        }
+        return res;
+    }
```

Comment:
I dislike the approach with ranges as json object names. Details below.

Reply:
me 2 do you have an alternative idea?